ccl/changefeedccl: TestChangefeedFailOnTableOffline failed

ccl/changefeedccl.TestChangefeedFailOnTableOffline [failed](https://tanzanite.cluster.engflow.com/invocations/default/266997f1-f50e-4209-aeac-fa624c0ee096?testReportRun=3&testReportShard=5&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on release-24.2 @ [8fd5bb201dbb7699d3c6231f52214b9d1884b314](https://github.com/cockroachdb/cockroach/commits/8fd5bb201dbb7699d3c6231f52214b9d1884b314):


```
=== RUN   TestChangefeedFailOnTableOffline
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestChangefeedFailOnTableOffline3713681840
    test_log_scope.go:76: use -show-logs to present logs inline
    changefeed_test.go:3987: -- test log scope end --
test logs left over in: outputs.zip/logTestChangefeedFailOnTableOffline3713681840
--- FAIL: TestChangefeedFailOnTableOffline (37.09s)
=== RUN   TestChangefeedFailOnTableOffline/cloudstorage/reverted_import_fails_changefeed_with_earlier_cursor
    helpers_test.go:869: making server as secondary tenant
    helpers_test.go:948: making cloudstorage feed factory
    sql_runner.go:237: expected query 'IMPORT INTO for_import CSV DATA ($1);' error 'pause point', got: pq: aborted in DistSender: result is ambiguous: context canceled
    --- FAIL: TestChangefeedFailOnTableOffline/cloudstorage/reverted_import_fails_changefeed_with_earlier_cursor (18.85s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=5</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedFailOnTableOffline.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-48132

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -5195,7 +5195,7 @@ func TestChangefeedFailOnTableOffline(t *testing.T) {
 		// Start an import job which will immediately pause after ingestion
 		sqlDB.Exec(t, "SET CLUSTER SETTING jobs.debug.pausepoints = 'import.after_ingest';")
 		go func() {
-			sqlDB.ExpectErrWithTimeout(t, `pause point`, `IMPORT INTO for_import CSV DATA ($1);`, dataSrv.URL)
+			sqlDB.ExpectErrWithRetry(t, `pause point`, `IMPORT INTO for_import CSV DATA ($1);`, `result is ambiguous`, dataSrv.URL)
 		}()
 		sqlDB.CheckQueryResultsRetry(
 			t,

--- a/pkg/testutils/sqlutils/BUILD.bazel
+++ b/pkg/testutils/sqlutils/BUILD.bazel
@@ -29,6 +29,7 @@ go_library(
         "//pkg/sql/sem/tree",
         "//pkg/testutils",
         "//pkg/util/protoutil",
+        "//pkg/util/retry",
         "//pkg/util/timeutil",
         "@com_github_cockroachdb_cockroach_go_v2//crdb",
         "@com_github_cockroachdb_errors//:errors",

--- a/pkg/testutils/sqlutils/sql_runner.go
+++ b/pkg/testutils/sqlutils/sql_runner.go
@@ -16,6 +16,7 @@ import (
 
 	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/errors"
 	"github.com/lib/pq"
@@ -242,6 +243,24 @@ func (sr *SQLRunner) ExpectErrWithTimeout(
 	if err != nil {
 		t.Fatalf("failed assert error: %s", err)
 	}
+}
+
+// ExpectErrWithRetry wraps ExpectErr with a timeout and retry on a specific error.
+func (sr *SQLRunner) ExpectErrWithRetry(
+	t Fataler, errRE string, query string, retryableErrorRE string, args ...interface{},
+) {
+	helperOrNoop(t)()
+	var err error
+
+	retryOpts := retry.Options{MaxRetries: 5}
+	err = retryOpts.DoWithRetryable(context.Background(), func(ctx context.Context) (bool, error) {
+		_, err = sr.DB.ExecContext(context.Background(), query, args...)
+		if testutils.IsError(err, retryableErrorRE) {
+			return true, err
+		}
+		return false, err
+	})
+	sr.expectErr(t, query, err, errRE)
 }
 
 // Query is a wrapper around gosql.Query that kills the test on error.

--- a/pkg/util/retry/retry.go
+++ b/pkg/util/retry/retry.go
@@ -180,6 +180,29 @@ func (opts Options) Do(ctx context.Context, fn func(ctx context.Context) error) 
 	return err
 }
 
+// DoWithRetryable invokes the closure according to the retry options until it
+// returns success or a non-retryable error. Always returns an error unless the
+// return is prompted by a successful invocation of `fn`.
+func (opts Options) DoWithRetryable(
+	ctx context.Context, fn func(ctx context.Context) (retryable bool, err error),
+) error {
+	var err error
+	for r := StartWithCtx(ctx, opts); r.Next(); {
+		var retryable bool
+		retryable, err = fn(ctx)
+		if err == nil {
+			return nil
+		}
+		if !retryable {
+			return err
+		}
+	}
+	if err == nil {
+		return errors.AssertionFailedf("never invoked function in DoWithRetryable")
+	}
+	return err
+}
+
 // WithMaxAttempts is a helper that runs fn N times and collects the last err.
 // The function will terminate early if the provided context is canceled, but it
 // guarantees that fn will run at least once.

